ccmbioinfo/lfs
# CARDS-2022: Add support for a Docker Compose environment with a reverse proxy that listens on two ports, one for administrative operations, one for user operations

This PR implements CARDS-2022 by adding the `--web_port_admin` and `--web_port_user` flags to `generate_compose_yaml.py` so that one port (`web_port_admin`) can be used by internal network staff for tasks that require a user account login (such as searching through data) while the other port (`web_port_user`) can be used for tasks that do not require a user account login (such as a patient on the public internet completing surveys).

Testing
---------

1. Build this (`CARDS-2022`) branch, including the Docker image with `mvn clean install -Pdocker`.
2. Download the `samlKeystore.p12` file from the `#cards-dev` Slack channel (it is pinned).
3. `cd compose-cluster`
4. Ensure that no Docker containers or volumes are present (`docker ps -a` and `docker volume ls` should both return empty lists).
5. Clean up any previous Docker Compose configurations with `./cleanup.sh`.

Testing HTTP Proxy Configuration
---------------------------------------------

1. Run `./quick-environments/proxy-ci-http.sh` and wait for it to start.
2. In a new terminal, enter the `Utilities/Development` directory and run `python3 test_proxy_configuration.py`, ensure that all tests pass.
3. You should also be able to visit http://localhost:8080 and interact with CARDS as normal.
4. Bring everything down by running, in the `compose-cluster` directory, `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.

Testing HTTP+SAML Proxy Configuration
-------------------------------------------------------

1. Run `./quick-environments/proxy-ci-http-saml-cloud-iam.sh` and wait for it to start.
2. In a new terminal, enter the `Utilities/Development` directory and run `python3 test_proxy_configuration.py --saml`, ensure that all tests pass.
3. You should also be able to visit http://localhost:8080 and interact with CARDS as normal. This includes being able to login to CARDS as `alice@cloud-iam.demo` or `bob@cloud-iam.demo`.
4. Bring everything down by running, in the `compose-cluster` directory, `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.

Testing HTTPS Proxy Configuration
------------------------------------------------

1. Run `./quick-environments/proxy-ci-https.sh` and wait for it to start.
2. In a new terminal, enter the `Utilities/Development` directory and run `python3 test_proxy_configuration.py --https`, ensure that all tests pass.
3. You should also be able to visit https://localhost (you'll need to accept the self-signed certificate) and interact with CARDS as normal.
4. Bring everything down by running, in the `compose-cluster` directory, `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.

Testing HTTPS+SAML Proxy Configuration
----------------------------------------------------------

1. Run `./quick-environments/proxy-ci-https-saml-cloud-iam.sh` and wait for it to start.
2. In a new terminal, enter the `Utilities/Development` directory and run `python3 test_proxy_configuration.py --saml --https`, ensure that all tests pass.
3. You should also be able to visit https://localhost (you'll need to accept the self-signed certificate) and interact with CARDS as normal. This includes being able to login to CARDS as `alice@cloud-iam.demo` or `bob@cloud-iam.demo`.
4. Bring everything down by running, in the `compose-cluster` directory, `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`.
